### PR TITLE
Stop sending messages on legacy substream altogether

### PR DESCRIPTION
**Pull request by tomaka**
Another step towards https://github.com/paritytech/substrate/issues/5670

This PR removes the legacy substream fallback for notification protocols, and the code that answers block requests on the legacy substream.
On top of this, I removed all the now-unused methods, which turned out to include the `send_message` method of the legacy substream. This proves that we're not sending any message on the legacy substream anymore, and we're ready to make it optional.

Due to removing the code that answers block requests, this PR breaks compatibility with Polkadot 0.7.33 and below.
Right now, according to the telemetry server, 27 Kusama nodes are concerned by this breaking change. However, out of these 27 nodes, only 6 are actually in sync. The other 21 nodes are stalled and probably abandoned by their owner. I do think that breaking compatibility with these 6 nodes is ok.
The Polkadot network isn't concerned, as we launched with Polkadot 0.8.0.

# Context

After a connection is established and a peerset slot reserved, we try to open the legacy substream plus one additional substream per notification protocol.

Before https://github.com/paritytech/substrate/pull/6821, it was possible for the legacy substream to be opened ahead of time compared to the other substreams. When that happened, we would automatically fall back to using the legacy substream in order to send notifications until the notification substreams are opened.
Now that https://github.com/paritytech/substrate/pull/6821 is merged, this cannot happen anymore. Notifications can only be sent or received when the corresponding (non-legacy) notification substream is open.

Apart from this fallback system, the legacy substream was also still used in order to receive block requests sent by nodes on Polkadot 0.7.33 and below. This PR removes this functionality. As such, this PR breaks compatibility with Polkadot 0.7.33 and below, as highlighted above.

### CODE REVIEW

**Comment by tomaka** [comment]
I've started a burn-in on my node: http://34.71.135.129:9090/graph?g0.range_input=6h&g0.expr=polkadot_sync_peers&g0.tab=0

**Comment by tomaka** [comment]
While this PR isn't really urgent by itself, we need to speed up a bit with the legacy substream removal in order to allow for a proper management of collators. As long as the legacy substream is there, we can't open a substream for a specific notification protocol without opening the substreams of all the other protocols.

**Comment by tomaka** [comment]
Considering that it potentially affects block production, we should deploy this on a Westend validator.

**Comment by tomaka** [comment]
I'm investigating the Polkadot test. Apparently we're failing to open a notification substream, and therefore using the legacy substream as a fallback as a result.

**Comment by tomaka** [comment]
I've found the issue and while open another PR. Putting this on ice in the meanwhile.

**Comment by tomaka** [comment]
I've merged #6979 in this PR.
Will un-ice this PR once #6979 is merged.

**Comment by tomaka** [comment]
Update: I think that the logic is now correct for actual nodes, but the GrandPa tests are failing because of the race conditions caused by #6827.

File: client/network/src/protocol/generic_proto/handler/group.rs [hunk 1]
```diff
@@ -602,26 +572,38 @@ impl ProtocolsHandler for NotifsHandler {
...
+									handler.send_or_discard(message);
+									continue 'poll_notifs_sink;
+								}
 							}
 						}
 
-						self.legacy.inject_event(LegacyProtoHandlerIn::SendCustomMessage {
-							message: encoded_fallback_message,
-						});
+						// If this code is reached, there exists two possibilities:
```
Comment (mxinden) [inline client/network/src/protocol/generic_proto/handler/group.rs:591]:
```suggestion
						// This code can be reached via the following scenarios:
```

**Review by mxinden — COMMENTED** [review]
Sorry for the delay. I will take a more in-depth look tomorrow.

File: client/finality-grandpa/src/tests.rs [hunk 2]
```diff
@@ -94,6 +94,15 @@ impl TestNetFactory for GrandpaTestNet {
 		ProtocolConfig::default()
 	}
 
+	fn add_full_peer(&mut self) {
```
Comment (mxinden) [inline client/finality-grandpa/src/tests.rs:97]: For any other reviewers, this function is called [here](https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/tests.rs#L74) and shadows [`add_full_peer` here](https://github.com/paritytech/substrate/blob/master/client/network/test/src/lib.rs#L617).

**Review by mxinden — APPROVED** [review]
This looks good to me. In case the burnin shows no further issues, I think we should merge here.

**Comment by tomaka** [comment]
bot merge

**Comment by tomaka** [comment]
bot merge

**Comment by tomaka** [comment]
bot merge

**Comment by tomaka** [comment]
bot merge

**Pull request merged**